### PR TITLE
Add optional input values and test suite

### LINKED ISSUE #25: Add `optional` key to config values as mutually-exclusive with `default`
It is sometimes handy to use the presence of a scalar to trigger some conditional behavior. For example: a `radius`, which when set will add the flag `--radius` to the algorithm's command-line invocation.

This conflicts with the manifest's need to provide firm assurances to the gear as to the structure of data it will be provided. For example: many languages cannot distinguish between a set, falsey / zero value, an unset value, or a null value. This ambiguity would make it difficult or impossible to correctly process a manifest.

So: let's make it opt-in. The `optional` key marks that a value may or may not be present at all in the `config.json`. This is mutually-exclusive with `default`, since it would be ambiguous to resolve the two.

Caveat: Depending the interface a consumer uses, an optional string value may not be very helpful. For example: in the web interface, empty strings should be sent if not filled out. @lmperry and I discussed this and it seems fair that since the algorithm opted into this, it's reasonable to have it deal with that.

This value is always provided, and defaults to zero:

```json
"speed": {
	"type": "integer",
	"default": "0",
	"minimum": 0,
	"maximum": 3,
	"description": "How fast do you want the gear to run? Choose 0-3."
}
```

This value is only provided if set, including zero:

```json
"speed": {
	"type": "integer",
	"optional": true,
	"minimum": 0,
	"maximum": 3,
	"description": "How fast do you want the gear to run? Choose 0-3."
}
```

This value is (probably) always provided by a web interface, and provided if set by a CLI interface:

```json
"name": {
	"type": "string",
	"optional": true,
	"description": "What is your name?"
}
```

### PR DESCRIPTION
**Pull request by kofalt**
Closes #25, #29. This feature was the breaking point for not having at least a basic validation suite, so that is added as well. Basic but functional. Comments welcome.

----

Is this a semantic or operational change? If so:

* [x] Increment the version in spec/readme.md
* [x] After merge, tag the version and update the release page

### CODE REVIEW

File: examples/test.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		else:
+			print '[X]'
+			return
+
+	print '[ ]'
+	raise Exception('Test should fail but did not: ' + name)
+
+def test_folders():
+	base = path.dirname(path.abspath(__file__))
+	print base
```
Comment (ehlertjd) [inline examples/test.py]: https://pythonclock.org/

**Comment by lmperry** [comment]
Thanks. LGTM

File: spec/readme.md [hunk 2]
```diff
@@ -150,6 +150,8 @@ Like the inputs, you can add JSON schema constraints as desired. Please specify
 
 The example has named one config option, called "speed", which must be an integer between zero and three, and another called "coordinates", which must be a set of three floats.
 
+In some cases, a configuration option may not have a safe default, and it only makes sense to sometimes omit it entirely. If that is the case, specify `"optional": "true"` on that config key.
```
Comment (ehlertjd) [inline spec/readme.md]: `"optional": true`

**Review by lmperry — APPROVED** [review]

**Review by ehlertjd — CHANGES_REQUESTED** [review]
One minor doc correction

File: examples/test.py (hunk 1 above)
Comment (kofalt) [reply examples/test.py]: :man_facepalming:

File: spec/readme.md (hunk 2 above)
Comment (kofalt) [reply spec/readme.md]: Fixed

**Review by ehlertjd — APPROVED** [review]
LGTM!

**Pull request merged**